File: sudoku_generator.js
// sudoku_generator.js — JavaScript версия

const fs = require('fs');

class SudokuGenerator {
    constructor(difficulty = 'medium', seed = null) {
        this.size = 9;
        this.boxSize = 3;
        this.board = Array.from({ length: 9 }, () => Array(9).fill(0));
        this.solution = Array.from({ length: 9 }, () => Array(9).fill(0));
        this.cellsToRemove = difficulty === 'medium' ? 45 : difficulty === 'hard' ? 50 : 35;
        this.difficulty = difficulty;
        if (seed !== null) {
            this._seed = seed;
        }
    }

    _rand() {
        if (this._seed !== undefined) {
            this._seed = (this._seed * 9301 + 49297) % 233280;
            return this._seed / 233280;
        }
        return Math.random();
    }

    generate() {
        this._solve(this.board);
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                this.solution[i][j] = this.board[i][j];
            }
        }
        this._removeCells();
        return this.board;
    }

    _solve(board) {
        const empty = this._findEmpty(board);
        if (!empty) return true;
        const [row, col] = empty;
        const nums = this._shuffle([1, 2, 3, 4, 5, 6, 7, 8, 9]);
        for (const num of nums) {
            if (this._isValid(board, row, col, num)) {
                board[row][col] = num;
                if (this._solve(board)) return true;
                board[row][col] = 0;
            }
        }
        return false;
    }

    _findEmpty(board) {
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                if (board[i][j] === 0) return [i, j];
            }
        }
        return null;
    }

    _isValid(board, row, col, num) {
        for (let j = 0; j < this.size; j++) {
            if (board[row][j] === num) return false;
        }
        for (let i = 0; i < this.size; i++) {
            if (board[i][col] === num) return false;
        }
        const startRow = Math.floor(row / this.boxSize) * this.boxSize;
        const startCol = Math.floor(col / this.boxSize) * this.boxSize;
        for (let i = startRow; i < startRow + this.boxSize; i++) {
            for (let j = startCol; j < startCol + this.boxSize; j++) {
                if (board[i][j] === num) return false;
            }
        }
        return true;
    }

    _removeCells() {
        const cells = [];
        for (let i = 0; i < this.size; i++) {
            for (let j = 0; j < this.size; j++) {
                cells.push([i, j]);
            }
        }
        this._shuffle(cells);
        let removed = 0;
        for (const [row, col] of cells) {
            if (removed >= this.cellsToRemove) break;
            this.board[row][col] = 0;
            removed++;
        }
    }

    _shuffle(arr) {
        for (let i = arr.length - 1; i > 0; i--) {
            const j = Math.floor(this._rand() * (i + 1));
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
        return arr;
    }

    hasUniqueSolution() {
        const boardCopy = this.board.map(row => [...row]);
        let count = 0;
        const countSolutions = (board) => {
            const empty = this._findEmpty(board);
            if (!empty) return 1;
            const [row, col] = empty;
            let c = 0;
            for (let num = 1; num <= 9; num++) {
                if (this._isValid(board, row, col, num)) {
                    board[row][col] = num;
                    c += countSolutions(board);
                    if (c >= 2) { board[row][col] = 0; return c; }
                    board[row][col] = 0;
                }
            }
            return c;
        };
        count = countSolutions(boardCopy);
        return count === 1;
    }

    printBoard(showSolution = false) {
        const board = showSolution ? this.solution : this.board;
        console.log('┌───────┬───────┬───────┐');
        for (let i = 0; i < this.size; i++) {
            let line = '│';
            for (let j = 0; j < this.size; j++) {
                line += board[i][j] === 0 ? ' . ' : ` ${board[i][j]} `;
                if (j % 3 === 2 && j < this.size - 1) line += '│';
            }
            line += '│';
            console.log(line);
            if (i % 3 === 2 && i < this.size - 1) {
                console.log('├───────┼───────┼───────┤');
            }
        }
        console.log('└───────┴───────┴───────┘');
    }

    saveTXT(filename = 'sudoku_puzzle.txt') {
        let content = '';
        for (const row of this.board) {
            content += row.map(c => c === 0 ? '.' : c).join(' ') + '\n';
        }
        fs.writeFileSync(filename, content);
        console.log(`💾 Сохранено: ${filename}`);
    }

    saveJSON(filename = 'sudoku_puzzle.json') {
        const data = {
            puzzle: this.board,
            solution: this.solution,
            difficulty: this.difficulty,
            cells_removed: this.cellsToRemove
        };
        fs.writeFileSync(filename, JSON.stringify(data, null, 2));
        console.log(`💾 Сохранено: ${filename}`);
    }

    saveHTML(filename = 'sudoku_puzzle.html') {
        let html = `<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Sudoku</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
table { border-collapse: collapse; margin: 20px auto; }
td { width: 40px; height: 40px; text-align: center; font-size: 20px; border: 1px solid #444; }
td:nth-child(3n) { border-right: 3px solid #fff; }
tr:nth-child(3n) td { border-bottom: 3px solid #fff; }
.empty { color: #555; }
.given { color: #4fc3f7; }
</style>
</head>
<body>
<h1 style="text-align:center;">🧩 Sudoku (Medium)</h1>
<table>
`;
        for (let i = 0; i < this.size; i++) {
            html += '<tr>';
            for (let j = 0; j < this.size; j++) {
                const val = this.board[i][j];
                const cls = val === 0 ? 'empty' : 'given';
                const display = val === 0 ? '·' : String(val);
                html += `<td class="${cls}">${display}</td>`;
            }
            html += '</tr>';
        }
        html += `</table>
<p style="text-align:center;">💡 Решение доступно в JSON файле</p>
</body></html>`;
        fs.writeFileSync(filename, html);
        console.log(`💾 Сохранено: ${filename}`);
    }
}

function main() {
    console.log('🧩 Sudoku Generator (Medium) (JavaScript)');
    const gen = new SudokuGenerator('medium');
    gen.generate();

    console.log('\nГоловоломка:');
    gen.printBoard();

    const readline = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });
    readline.question('\nПоказать решение? (y/n): ', (ans) => {
        if (ans.toLowerCase() === 'y') {
            console.log('\nРешение:');
            gen.printBoard(true);
        }
        gen.saveTXT();
        gen.saveJSON();
        gen.saveHTML();
        readline.close();
    });
}

if (require.main === module) main();

File: sudoku_generator.go
// sudoku_generator.go — Go версия

package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"time"
)

type SudokuGenerator struct {
	Size          int
	BoxSize       int
	Board         [][]int
	Solution      [][]int
	CellsToRemove int
	Difficulty    string
	Seed          int64
}

func NewSudokuGenerator(difficulty string, seed int64) *SudokuGenerator {
	remove := 45
	if difficulty == "easy" {
		remove = 35
	} else if difficulty == "hard" {
		remove = 50
	}
	board := make([][]int, 9)
	solution := make([][]int, 9)
	for i := range board {
		board[i] = make([]int, 9)
		solution[i] = make([]int, 9)
	}
	return &SudokuGenerator{
		Size:          9,
		BoxSize:       3,
		Board:         board,
		Solution:      solution,
		CellsToRemove: remove,
		Difficulty:    difficulty,
		Seed:          seed,
	}
}

func (s *SudokuGenerator) generate() {
	if s.Seed != 0 {
		rand.Seed(s.Seed)
	} else {
		rand.Seed(time.Now().UnixNano())
	}
	s.solve(s.Board)
	for i := 0; i < s.Size; i++ {
		for j := 0; j < s.Size; j++ {
			s.Solution[i][j] = s.Board[i][j]
		}
	}
	s.removeCells()
}

func (s *SudokuGenerator) solve(board [][]int) bool {
	row, col, found := s.findEmpty(board)
	if !found {
		return true
	}
	nums := rand.Perm(9)
	for _, n := range nums {
		num := n + 1
		if s.isValid(board, row, col, num) {
			board[row][col] = num
			if s.solve(board) {
				return true
			}
			board[row][col] = 0
		}
	}
	return false
}

func (s *SudokuGenerator) findEmpty(board [][]int) (int, int, bool) {
	for i := 0; i < s.Size; i++ {
		for j := 0; j < s.Size; j++ {
			if board[i][j] == 0 {
				return i, j, true
			}
		}
	}
	return 0, 0, false
}

func (s *SudokuGenerator) isValid(board [][]int, row, col, num int) bool {
	for j := 0; j < s.Size; j++ {
		if board[row][j] == num {
			return false
		}
	}
	for i := 0; i < s.Size; i++ {
		if board[i][col] == num {
			return false
		}
	}
	startRow := (row / s.BoxSize) * s.BoxSize
	startCol := (col / s.BoxSize) * s.BoxSize
	for i := startRow; i < startRow+s.BoxSize; i++ {
		for j := startCol; j < startCol+s.BoxSize; j++ {
			if board[i][j] == num {
				return false
			}
		}
	}
	return true
}

func (s *SudokuGenerator) removeCells() {
	cells := make([][2]int, 0)
	for i := 0; i < s.Size; i++ {
		for j := 0; j < s.Size; j++ {
			cells = append(cells, [2]int{i, j})
		}
	}
	rand.Shuffle(len(cells), func(i, j int) { cells[i], cells[j] = cells[j], cells[i] })
	removed := 0
	for _, cell := range cells {
		if removed >= s.CellsToRemove {
			break
		}
		s.Board[cell[0]][cell[1]] = 0
		removed++
	}
}

func (s *SudokuGenerator) printBoard(showSolution bool) {
	board := s.Board
	if showSolution {
		board = s.Solution
	}
	fmt.Println("┌───────┬───────┬───────┐")
	for i := 0; i < s.Size; i++ {
		line := "│"
		for j := 0; j < s.Size; j++ {
			if board[i][j] == 0 {
				line += " . "
			} else {
				line += fmt.Sprintf(" %d ", board[i][j])
			}
			if j%3 == 2 && j < s.Size-1 {
				line += "│"
			}
		}
		line += "│"
		fmt.Println(line)
		if i%3 == 2 && i < s.Size-1 {
			fmt.Println("├───────┼───────┼───────┤")
		}
	}
	fmt.Println("└───────┴───────┴───────┘")
}

func (s *SudokuGenerator) saveTXT(filename string) {
	f, _ := os.Create(filename)
	defer f.Close()
	for _, row := range s.Board {
		for j, cell := range row {
			if cell == 0 {
				f.WriteString(". ")
			} else {
				f.WriteString(strconv.Itoa(cell) + " ")
			}
			if j == 8 {
				f.WriteString("\n")
			}
		}
	}
	fmt.Printf("💾 Сохранено: %s\n", filename)
}

func (s *SudokuGenerator) saveJSON(filename string) {
	data := map[string]interface{}{
		"puzzle":         s.Board,
		"solution":       s.Solution,
		"difficulty":     s.Difficulty,
		"cells_removed":  s.CellsToRemove,
	}
	jsonData, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(filename, jsonData, 0644)
	fmt.Printf("💾 Сохранено: %s\n", filename)
}

func (s *SudokuGenerator) saveHTML(filename string) {
	html := `<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"><title>Sudoku</title>
<style>
body { font-family: monospace; background: #1a1a2e; color: #fff; padding: 20px; }
table { border-collapse: collapse; margin: 20px auto; }
td { width: 40px; height: 40px; text-align: center; font-size: 20px; border: 1px solid #444; }
td:nth-child(3n) { border-right: 3px solid #fff; }
tr:nth-child(3n) td { border-bottom: 3px solid #fff; }
.empty { color: #555; }
.given { color: #4fc3f7; }
</style>
</head>
<body>
<h1 style="text-align:center;">🧩 Sudoku (Medium)</h1>
<table>
`
	for i := 0; i < s.Size; i++ {
		html += "<tr>"
		for j := 0; j < s.Size; j++ {
			val := s.Board[i][j]
			cls := "given"
			if val == 0 {
				cls = "empty"
			}
			display := "·"
			if val != 0 {
				display = strconv.Itoa(val)
			}
			html += fmt.Sprintf("<td class='%s'>%s</td>", cls, display)
		}
		html += "</tr>"
	}
	html += `</table>
<p style="text-align:center;">💡 Решение доступно в JSON файле</p>
</body></html>`
	os.WriteFile(filename, []byte(html), 0644)
	fmt.Printf("💾 Сохранено: %s\n", filename)
}

func main() {
	fmt.Println("🧩 Sudoku Generator (Medium) (Go)")
	gen := NewSudokuGenerator("medium", 0)
	gen.generate()

	fmt.Println("\nГоловоломка:")
	gen.printBoard(false)

	fmt.Print("\nПоказать решение? (y/n): ")
	var ans string
	fmt.Scanln(&ans)
	if ans == "y" || ans == "Y" {
		fmt.Println("\nРешение:")
		gen.printBoard(true)
	}

	gen.saveTXT("sudoku_puzzle.txt")
	gen.saveJSON("sudoku_puzzle.json")
	gen.saveHTML("sudoku_puzzle.html")
}
